### ISSUE TITLE
btrfs-receive slows down exponentially (on large transfers)

### ISSUE THREAD

**Issue by AmNotOnline** [kdave/btrfs-progs#908]
I'm in the process of migrating my personal backups from an rsync-based approach to use snapshots. This means that I have to bootstrap a snapshot on my storage server. My home subvolume measured about 550 GiB (as reported by `du -sh`), and dumping that to a file via `btrfs send --compressed-data -f ...` resulted in a file of 419 GiB.

Initially, I tried sending the bootstrap snapshot over via an ssh connection, which more or less saturated the write speed of the receiving 1 TB HDD drive dedicated solely to backups. Yet, after multiple hours, this slows down to less than a megabyte transferred per second.

I tried this approach three times on separate days, completely wiping the receiving drive between transfer attempts.

Then, I tried dumping the snapshot to a file on an external HDD drive, and manually receiving it into my backup server. Again, for the first few hundred gigabytes of the transfer, the transfer speed was limited by the physical limitations of the drives. Eventually, this dropped to 600-800 KB/s (as reported by `btop` as total drive access for the backup drive).

This leads me to the conclusion that the problem is with `btrfs receive`, and not my flaky LAN setup.

**TL;DR**
Large snapshot transfers using `btrfs receive` slow down exponentially when far into large data loads.

**Receiving System:**
- CPU: Intel Duo E7500
- RAM: 4 GB
- Backup Drive: Seagate Mobile ST1000LM035-1RK172 (1 TB)

**Additional info:**
- Ran tests with `smartctl` on receiving drive, no errors were reported.
- CPU usage only went as high as 5 %, so that shouldn't be the bottleneck.
- Another 40 GB transfer using `btrfs receive -f` worked totally fine.

**Comment by joshhansen** [comment]
I think I'm seeing something similar, but it originates from the send side, since the slowdown happens even when I pipe the send output to /dev/null instead of sending it over ssh.

The transfer is a 20TB subvolume. It starts seemingly at full speed with 232 MiB/s average over 30 minutes. But by the end, at times the instant transfer speed drops to 0, and the average to single-digit megabytes / s. I'm also using `--compressed-data`.

Kernel 6.12.57
Btrfs progs 6.14

4 SATA drives in raid1 config for data and metadata.

(This is on current Debian stable.)

Will try again on a more recent kernel soon.
